### PR TITLE
RR-580 - Map from prison-api model into PrisonMovementEvents

### PR DESCRIPTION
**Pull request by mattwills-moj-digital**
This PR converts the `PrisonerInPrisonSummary` objects we receive from the `prison-api` into `PrisonerInPrisonSummary`, which are returned from our `PrisonApiClient`.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    prisonerNumber: String,
+    prisonerInPrisonSummary: PrisonerInPrisonSummary,
+  ): PrisonMovementEvents {
+    val prisonMovements = mutableMapOf<Long, List<PrisonMovementEvent>>()
+
+    prisonerInPrisonSummary.prisonPeriod?.forEach {
+      val movementEvents = mutableListOf<PrisonMovementEvent>()
+      movementEvents.addAll(buildArrivalsAndReleases(it.movementDates))
+      movementEvents.addAll(buildTransfers(it.transfers))
+      prisonMovements[it.bookingId!!] = movementEvents
```
Comment (mattwills-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt:20]: I will double check with Andy Marke whether `bookingId` will ever be null. If not, I'll create a PR for this spec.

File: src/main/resources/static/openapi/PrisonApi.json [hunk 2]
```diff
@@ -28940,12 +28940,14 @@
           "entryDate": {
             "pattern": "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}$",
             "type": "string",
+            "format": "date-time",
```
Comment (mattwills-moj-digital) [inline src/main/resources/static/openapi/PrisonApi.json:28943]: I tried to find out how to add this to the relevant `@Schema` annotations within the prison-api (e.g. `format = "date-time"`), but I've drawn a blank so far.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt (hunk 1 above)
Comment (nathanrussell-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt:20]: If it could be null we might need to rethink our data structure re: creating a map of movements keyed on bookingId 🤔

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    return PrisonMovementEvents(prisonerNumber, prisonMovements)
+  }
+
+  private fun buildArrivalsAndReleases(movements: List<SignificantMovement>): List<PrisonMovementEvent> {
+    val arrivalsAndReleases = mutableListOf<PrisonMovementEvent>()
+    movements.forEach {
+      if (isAdmissionIntoPrison(it)) {
+        arrivalsAndReleases.add(toPrisonArrivalEvent(it))
```
Comment (nathanrussell-moj-digital) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt:30]: Minor, but we might want to check with someone like Andy Marke what the official terminology is re: admissions.
They appear to have called it admissions (ADM), but we are using the term arrival ... I wonder if there's any mileage in being more aligned in this respect and it's potentially some key HMPPS terminology ?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapperTest.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    // Then
+    assertThat(actual).isEqualTo(expected)
+  }
+
+  @Test
+  fun `should convert to PrisonMovementEvents given empty prison periods`() {
+    // Given
+    val prisonNumber = aValidPrisonNumber()
+    val prisonSummary = aValidPrisonerInPrisonSummary(
+      prisonPeriod = emptyList(), // not expected to happen in practice - there should always be an arrival event(?)
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapperTest.kt:51]: 100% agreed, but its right that we test edge cases like this 👍

File: src/main/resources/static/openapi/PrisonApi.json (hunk 2 above)
Comment (mattwills-moj-digital) [reply src/main/resources/static/openapi/PrisonApi.json:28943]: I'm not comfortable changing our copy of their spec either, but at least if someone overwrites it then the build will fail. They'd then have to change the corresponding builder and mapper etc, so hopefully that will be enough to make them stop in their tracks! 😄

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt (hunk 3 above)
Comment (mattwills-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt:30]: I'll use `Arrivals` instead, as that's what they've called it (at least most of the time - I've found `Arrivals` in their spec too!).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt (hunk 1 above)
Comment (mattwills-moj-digital) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/client/prisonapi/PrisonApiMapper.kt:20]: I made the field mandatory in the prison-api: https://github.com/ministryofjustice/prison-api/pull/1988

**Review by nathanrussell-moj-digital — APPROVED** [review]
As usual, a thing of absolute beauty and delight 😁 
✅
